coderedcms/models/page_models.py

```diff
@@ -705,6 +716,189 @@ class Meta:
+
+            # Add the qualifying generated event instances to the list.
+            event_instances += [instance for instance in occurrence.all_occurrences(**occurrence_kwargs)]
+
+        # Sort all the events by the date that they start
+        event_instances.sort(key=lambda d: d[0])
+
+        # Return the event instances, possibly spliced if num_instances_to_return is set.
+        return event_instances[:num_of_instances_to_return] if num_of_instances_to_return else event_instances
+
+    def convert_to_ical_format(self, dt_start=None, dt_end=None, occurrence=None):
+        ical_event = ICalEvent()
+        ical_event.add('summary', self.title)
+        if dt_start:
+            ical_event.add('dtstart', dt_start)
+
+            if dt_end:
+                ical_event.add('dtend', dt_end)
+
+        if occurrence:
+            freq = occurrence.repeat.split(":")[1] if occurrence.repeat else None
+            repeat_until = occurrence.repeat_until.strftime("%Y%m%dT000000Z") if occurrence.repeat_until else None
+
+            ical_event.add('dtstart', occurrence.start)
+
+            if occurrence.end:
+                ical_event.add('dtend', occurrence.end)
+
+            if freq:
+                ical_event.add('RRULE', freq, encode=False)
+
+            if repeat_until:
+                ical_event.add('until', repeat_until)
+
+        return ical_event
+
+    def create_single_ical(self, dt_start, dt_end=None):
+        return self.convert_to_ical_format(dt_start=dt_start, dt_end=dt_end)
+
+    def create_recurring_ical(self):
+        events = []
+        for occurrence in self.occurrences.all():
+            events.append(self.convert_to_ical_format(occurrence=occurrence))
+        return events
+
+class DefaultCalendarViewChoices():
+    MONTH = 'month'
+    AGENDA_WEEK = 'agendaWeek'
+    AGENDA_DAY = 'agendaDay'
+    LIST_MONTH = 'listMonth'
+
+    CHOICES = (
+            (MONTH, 'Monthly Calendar'),
+            (AGENDA_WEEK, 'Weekly Calendar'),
+            (AGENDA_DAY, 'Daily Calendar'),
+            (LIST_MONTH, 'Monthly List'),
+        )
+
+class CoderedEventIndexPage(CoderedWebPage):
+    """
+    Shows a list of event sub-pages.
+    """
+    class Meta:
+        verbose_name = _('CodeRed Event Index Page')
+        abstract = True
+
+    template = 'coderedcms/pages/event_index_page.html'
+
+    index_show_subpages_default = True
+
+    NEXT_OCCURRENCE_ATTR = 'next_occurrence'
+
+    default_calendar_view = models.CharField(
+        blank=True,
+        choices=DefaultCalendarViewChoices.CHOICES,
+        max_length=255,
+        verbose_name=_('Default Calendar View'),
+        help_text=_('The default look of the calendar on this page.')
+    )
+
+    layout_panels = (
+        CoderedWebPage.layout_panels +
+        [
+            FieldPanel('default_calendar_view'),
+        ]
+    )
+
+    def get_index_children(self):
+        if self.index_query_pagemodel and self.index_order_by == self.NEXT_OCCURRENCE_ATTR:
```

Comment:
I did this to remove a hard coded value from function level code.  `NEXT_OCCURRENCE_ATTR` is a constant and shouldn't be changed.

Reply:
Yes, but it has to be set to `'next_occurrence'` otherwise it will break. So I think it should be *more* hard-coded. The overriding in the website/models.py makes it look like it's something that should be changed. We could probably actually move this code completely back into the abstract model:

```python
# Override to specify custom index ordering choice/default. 
NEXT_OCCURRENCE_ATTR = 'next_occurrence' 
index_query_pagemodel = 'website.EventPage' 
index_order_by_default = NEXT_OCCURRENCE_ATTR 
index_order_by_choices = ( 
    (NEXT_OCCURRENCE_ATTR, 'Display next occurrence, soonest first'), 
) + \ 
CoderedEventIndexPage.index_order_by_choices 
```

Reply:
There's similar logic in the project template for the `ArticleIndexPage`.  Should that be moved too?

Reply:
Well, I guess it needs to be there in website/models.py. But it could probably be changed to:
```python
# Override to specify custom index ordering choice/default. 
index_query_pagemodel = 'website.EventPage' 
index_order_by_default = 'next_occurrence'
index_order_by_choices = ( 
    ('next_occurrence',  'Display next occurrence, soonest first'), 
) + \ 
CoderedEventIndexPage.index_order_by_choices 
```

and then page_models.py could be changed to remove `NEXT_OCCURRENCE_ATTR` and replace it in the function with a hard-coded value. Because it is not dynamic, it is literally a function. It could be defined as a variable inside the `get_index_children()` method, but should be not changeable outside that method otherwise it would break that method. Does that make sense?

We would then document that `'next_occurence'` is a special value that can be passed into `index_order_by_choices` when the `index_query_pagemodel` inherits from `CoderedEventPage`.

Reply:
The only thing that actually needs to be in `website/models.py` is the `index_query_pagemodel`.  The rest is fine to be moved into the abstract.  That goes for both the `ArticleIndexPage` and the `EventIndexPage`.